lnst/RecipeCommon/Perf/Measurements/TcRunMeasurement.py

```diff
@@ -0,0 +1,244 @@
+from lnst.RecipeCommon.Perf.Measurements.Results.AggregatedTcRunMeasurementResults import AggregatedTcRunMeasurementResults
+from lnst.RecipeCommon.Perf.Measurements.Results.TcRunMeasurementResults import TcRunMeasurementResults
+from lnst.RecipeCommon.Perf.Results import PerfInterval
+from lnst.Tests.TrafficControl import TrafficControlRunner
+from lnst.Controller.Job import Job
+from lnst.Controller.Namespace import Device, Namespace
+from lnst.RecipeCommon.Perf.Measurements.BaseMeasurement import BaseMeasurement
+
+
+class BaseTimedRunMeasurement(BaseMeasurement):
+    # TODO Are there useful things to put into this subclass
+    # For other potential programs where we want to measure the
+    # amount of time something took.
+    pass
+
+
+class TcRunInstance:
+    MAX_ID = 0xFFFF
+
+    def __init__(self, device: Device, num_rules: int, instance_id:int):
+        self._device = device
+        self._num_rules = num_rules
+        self._instance_id = instance_id
+        self._batchfile_path: PathLike = None
+        self._instance_job: Job = None
+        self.validate()
+
+    @property
+    def instance_job(self):
+        return self._instance_job
+
+    @instance_job.setter
+    def instance_job(self, job: Job):
+        self._instance_job = job
+
+
+    @property
+    def batchfile_path(self):
+        return self._batchfile_path
+
+    @batchfile_path.setter
+    def batchfile_path(self, p: PathLike):
+        self._batchfile_path = p
+
+    def validate(self):
+        if self._instance_id > self.MAX_ID:
+            raise ValueError(f"Maximum number of instances supported is {self.MAX_ID}")
+        try:
+            required_macs = self._num_rules * 2 # Each rule needs src/dest mac
+            required_macs.to_bytes(3,'big')
+        except OverflowError:
+            raise ValueError(f"Can't create enough MAC addresses for {self._num_rules}")
+
+    @property
+    def num_rules(self) -> int:
+        return self._num_rules
+
+    @property
+    def pool_oui(self):
+        # Somewhat simple way to create a mac pool of unicast address
+        return f"00:{self._instance_id.to_bytes(2, 'big').hex(':')}"
+
+    @property
+    def device(self) -> Device:
+        return self._device
+
+    @property
+    def host(self) -> Namespace:
+        return self._device.host
+
+    @property
+    def start_mac(self) -> str:
+        return f"{self.pool_oui}:00:00:00"
+
+    @property
+    def end_mac(self) -> str:
+        return f"{self.pool_oui}:ff:ff:ff"
+
+    def generate_batchfile(self) -> str:
+        logging.info(f"Generating tc batchfile for {self.device.name} instance {self._instance_id} num_rules={self.num_rules}")
+        rules = self.generate_rules()
+        with NamedTemporaryFile(
+                'w', suffix=".batch", prefix="tc-rules-", delete=True,
+        ) as f:
+            for r in rules:
+                f.write(r)
+
+            remote_path = self.host.copy_file_to_machine(f.name)
```

Comment:
I was doing that before. I changed it for a few difference reasons.  

I need to have the batchfile created and ready to go before running the job.  Generating these files (100k lines) takes a not long, but not insignificant amount of time

We don't seem to have a standard way of doing "pre-start" tasks that would get run when calling `host.prepare_job`. I guess I could override `__init__` and do that. Alternatively I could create another "test file creator job". I considered some of those when refactoring this, but decided to go with this approach because i felt it was cleaner. 

Its possible that there are errors during creation, I feel those are easier to see, handle, and debug during development if they happen on the controller. 

I feel like this way unifies it more and max it more flexible, in the future if we find issues with it should be easy to move it to the agent. Perhaps I could move the `generate_rules` function itself to some common class so it could be used either way.

Reply:
ok, that makes sense.